llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -464,7 +467,7 @@ struct TernaryOpc_match {
   bool match(const MatchContext &Ctx, SDValue N) {
     if (sd_context_match(N, Ctx, m_Opc(Opcode))) {
       EffectiveOperands<ExcludeChain> EO(N);
-      assert(EO.Size == 3);
+      assert(EO.Size == 3U + 2 * N->isVPOpcode());
```

Comment:
**mshockwave**: https://github.com/llvm/llvm-project/pull/102877#discussion_r1715574396

> would it be better if we put this into `EffectiveOperands`? `EffectiveOperands` was invented to abstract away "don't care" operands, like input chain, under certain circumstances so I feel like it fits well with what you're doing here.

Reply:
We could add another template argument `bool ExcludeVPArgs`. When set to `true` and a VP Node, `TotalNumOperands := N->getNumOperands() - 2`.

It supposes the implicit convention that with VP nodes the last two args are for mask and vector length. Do you want me to search if it is always true and write it down here ?

Reply:
Well actually, I added a Ctx argument to the constructor instead.